### PR TITLE
JETC-861: Get SSPH working on Python 3

### PR DESCRIPTION
**Pull request by ariedel**
This PR contains a simple conversion to Python 3. ~There wasn't much additional refactoring done.~
This PR includes a conversion to Python 3 and a light refactoring.
* Removed our dependency on ipaddr ([obsoleted](https://pypi.org/project/ipaddr/) by the standard library ipaddress function), 
* Removed our dependency on iso8601 ([obsoleted](https://pypi.org/project/iso8601/) by the dateutil package, which is even recommended by python's own [documentation](https://docs.python.org/3/library/datetime.html#module-datetime) on datetime)
* Removed our dependency on pytz in favor of dateutil.
* More pythonic: Changed all `if 0` switch statements to `if True`, and all `return (number)` to `sys.exit()` commands
* `hash` appears to be a reserved word, so I changed the variable name
* The `exec()` function behaves differently in python 3, so I've replaced it where it was used to generate the hash.

what has been tested as working:

- On the admin page:
  - duplicating the window (restarting the CGI, basically)
  - List SP
  - List Auths
  - Adding an SP with the form fields (this was working earlier, I just misunderstood what went in the dbcreds field)
  - Get password
  - Remove an SP by name

* Other functionality:
  * Logging in to a test slot

 

What doesn't work:

* Setting the password via the admin page (because it's not writeable... nor do we need it to be)

### CODE REVIEW

File: ssph_server/db.py [hunk 1]
```diff
@@ -19,25 +19,25 @@
...
 # database: sqlite
 # MOVE THE DATABASE LOCATION BEFORE ACTUAL USE
-if 0:
+if True:
     import pandokia.db_sqlite as d
     # sqlite only needs a file, but it needs read/write on the file and
     # the directory it is in.  sqlite is not very good at handling lots
     # of concurrent transactions.
-    core_db = d.PandokiaDB("/tmp/test_ssph.db")
+    core_db = d.PandokiaDB("/internal/data1/other/logs/test_db.db")
```
Comment (ariedel) [inline ssph_server/db.py]: This absolutely needs to be changed back to the mysql database before merging

File: ssph_server/db.py [hunk 2]
```diff
@@ -19,25 +19,25 @@
...
+try:
+    password = open(password_file, "r").readline().strip()
+except IOError:
     password = None
 
 #####
 # database: sqlite
 # MOVE THE DATABASE LOCATION BEFORE ACTUAL USE
-if 0:
+if True:
```
Comment (dpnemergut) [inline ssph_server/db.py]: Flipped logic here?

File: ssph_server/db.py [hunk 3]
```diff
@@ -51,7 +51,7 @@
 #####
 # database: mysql
 
-if 1:
+if False:
```
Comment (dpnemergut) [inline ssph_server/db.py]: Mean to flip this logic?

File: ssph_server/db.py (hunk 2 above)
Comment (ariedel) [reply ssph_server/db.py]: Yes. For testing purposes, I'm using a sqlite3 database, so I activated it here...

File: ssph_server/db.py (hunk 3 above)
Comment (ariedel) [reply ssph_server/db.py]: ...and deactivated mysql here.

Comment (dpnemergut) [reply ssph_server/db.py]: I see it now. More confusion about why we use sqlite and not mysql in local/pandokia but that's a conversation for another, much more distant, day.

Comment (ariedel) [reply ssph_server/db.py]: This is a temporary change just for debugging purposes. We normally use a mysql database for this.

**Comment by ariedel** [comment]
I have now added the CGI scripts to the repo.

**Comment by ariedel** [comment]
Code now points to the MySQL database

**Comment by ariedel** [comment]
Ok, I am a little worried about bytes->str issues here; when I view the SSPH database from plssph4 (in the admin page) all of the fields (except auth_event_id, which is bytes) are strings. When I do the same for our production database, all of the fields are bytes. This is processing by pandokia's text_table() function, so it would seem to be a difference.

I see two bad outcomes here if I actually point this at the production database:
1.) the event IDs will be misprocessed when they're retrieved and won't authenticate properly (which might be a minor concern, because the servers shouldn't be authenticating a connection that was started on a different server.
2.) the existing database might require data to be specified as bytes. It has tables that were created with bytes objects (well, non-unicode strings) and we know MySQL is picky about that.

The best way to test this, ultimately, would be to ask ITSD to make a backup of the production SSPH database and host it somewhere where I can test against THAT. That's a lot of additional time to invest.

**Comment by ariedel** [comment]
The path to getting both SSPH servers updated to this version seems clear:
1. Convert code to Python 3
2. Get ITSD to make a copy of the production plssphdb2 database on another system and point plssph4 at it
3. Debug any problems with the database
4. Switch plssph4 to pointing at the production plssphdb2 database
5. Roll out code to plssph3

Step 1 is done, and may technically fulfil the sprint goal. Step 5 definitely can't be done before the call closes.

The question (for the team, especially @vglaidler) is do we do steps 2-4 now? I don't know how likely it would be to get step 2 done by the end of the sprint.

**Comment by oiintam** [comment]
So I don't really know if I am fully understanding the bytes to string situation. If you are saying that the fields in the ssph database tables themselves are bytes instead of string, I believe you can change the type of the field from bytes to string and then we can enforce the future entries to be a certain type. And of course we will test carefully on a copy of production first.

**Comment by oiintam** [comment]
Also, I can help you with the conversion of the database if you would like to.

**Comment by ariedel** [comment]
With the help of Jeff Wagner, I have now accomplished steps 2-4. 
2. Jeff Wagner plssph4 made a copy of the current plssphdb2 ssph database and hosted it on dlgreenmysql
3. I pointed plssph4 at that server and determined that it now needs to be loaded with unicode=True; three people were able to log in to test_3 while it was using plssph4+dlgreenmysql
4. The change is now made and plssph4 is now pointing to plssphdb2. I have successfully logged in.

**Comment by ariedel** [comment]
I am now satisfied the SSPH conversion work is completely finished apart from actually deploying it to plssph3/ssph2.etc.stsci.edu

**Pull request merged**